components/script/dom/bindings/js.rs

```diff
@@ -391,6 +392,86 @@ impl<T: DomObject> HeapSizeOf for MutNullableJS<T> {
     }
 }
 
+/// A holder that allows to lazily initialize the value only once
+/// `JS<T>`, using OnceCell
+/// Essentially a `OnceCell<JS<T>>`.
+///
+/// This should only be used as a field in other DOM objects; see warning
+/// on `JS<T>`.
+#[must_root]
+pub struct OnceCellJS<T: DomObject> {
+    ptr: OnceCell<JS<T>>,
+}
+
+impl<T: DomObject> OnceCellJS<T> {
+    /// Create a new `OnceCellJS` with value
+    pub fn new(initial: Option<&T>) -> OnceCellJS<T> {
+        debug_assert!(thread_state::get().is_script());
+        OnceCellJS {
+            ptr: OnceCell::new_with_value(initial.map(JS::from_ref).unwrap()),
+        }
+    }
+
+    /// Retrieve a copy of the current inner value. If it is `None`, it is
+    /// initialized with the result of `cb` first.
+    #[allow(unrooted_must_root)]
+    pub unsafe fn or_init<F>(&self, cb: F) -> &JS<T>
```

Comment:
I'm still not convinced this is an improvement over `MutNullableJS<T>`. In fact as it stands currently, it is an exact copy of it. Let me reiterate the goals of the issue:

Consider the [Elements method in HTMLFormElement](https://github.com/servo/servo/blob/74558990b258cb55f230ebe8ec3fc557fd286f94/components/script/dom/htmlformelement.rs#L170). It tries to see if the `elements` field is a `Some`; if so, it would return that inner value, and if not, it does a whole chunk of initialization for it, and finally setting that inner value into `Some`.

What I personally would like to see is to have that method body be reduced into simply `self.elements.get()`, and perhaps in the `new` method of the OnceCellJS that houses the list of elements, accept a closure/function that gets called whenever the `get` method encounters a `None` for its inner value.